meteor/server/api/ingest/generation.ts

```diff
@@ -128,11 +128,45 @@ export async function calculateSegmentsFromIngestData(
 				rundown,
 				watchedPackages
 			)
+			let blueprintSegment0: BlueprintResultSegment | null = null
+			try {
+				blueprintSegment0 = blueprint.blueprint.getSegment(context, ingestSegment)
+			} catch (err) {
+				logger.error(`Error in showStyleBlueprint.getSegment: ${stringifyError(err)}`)
+				blueprintSegment0 = null
+			}
 
-			const blueprintRes = blueprint.blueprint.getSegment(context, ingestSegment)
+			if (!blueprintSegment0) {
+				// Something went wrong when generating the segment
+
+				const newSegment = literal<DBSegment>({
+					_id: segmentId,
+					rundownId: rundown._id,
+					externalId: ingestSegment.externalId,
+					externalModified: ingestSegment.modified,
+					_rank: ingestSegment.rank,
+					notes: [
+						{
+							type: NoteSeverity.ERROR,
+							message: {
+								key: 'Internal Error generating segment',
+								namespaces: [unprotectString(blueprint.blueprintId)],
+							},
+							origin: {
+								name: '', // TODO
+							},
+						},
+					],
+					name: ingestSegment.externalId,
+				})
+				res.segments.push(newSegment)
+
+				continue // Don't generate any content for the faulty segment
```

Comment:
Is the correct behaviour here to remove all the faulty parts, or should we preserve the last 'good' copy of them?

I have a feeling that if this does succeed in removing all the old parts, then the we will set the segment as hidden and so it will disappear, which is probably not desirable. I dont know if the note will still get shown.

But maybe this is fine, as the blueprints could do proper error handling instead. In our blueprints getSegment already does this for each part generation, where if any part throws an error while generating it will be replaced by an invalid part

Reply:
I'm not 100%, but I'm leaning towards that this is fine.
I checked just now and the empty segment isn't hidden and the note shows so that's good.

As you say, this is protecting against an edge-case, if the blueprints handles themselves properly this'll never happen.